crates/ra_assists/src/handlers/fill_match_arms.rs

```diff
@@ -56,31 +50,115 @@ pub(crate) fn fill_match_arms(ctx: AssistCtx) -> Option<Assist> {
     }
 
     let db = ctx.db;
-
     ctx.add_assist(AssistId("fill_match_arms"), "Fill match arms", |edit| {
-        let indent_level = IndentLevel::from_node(match_arm_list.syntax());
+        let mut arms: Vec<MatchArm> = match_arm_list.arms().collect();
+        if arms.len() == 1 {
+            if let Some(Pat::PlaceholderPat(..)) = arms[0].pat() {
+                arms.clear();
+            }
+        }
 
-        let new_arm_list = {
-            let arms = variants
-                .into_iter()
-                .filter_map(|variant| build_pat(db, module, variant))
-                .map(|pat| make::match_arm(iter::once(pat), make::expr_unit()));
-            indent_level.increase_indent(make::match_arm_list(arms))
-        };
+        let mut has_partial_match = false;
+        let variants: Vec<MatchArm> = variants
+            .into_iter()
+            .filter_map(|variant| build_pat(db, module, variant))
+            .filter(|variant_pat| {
+                !arms.iter().filter_map(|arm| arm.pat().map(|_| arm)).any(|arm| {
+                    let pat = arm.pat().unwrap();
+
+                    // Special casee OrPat as separate top-level pats
+                    let pats: Vec<Pat> = match Pat::from(pat.clone()) {
+                        Pat::OrPat(pats) => pats.pats().collect::<Vec<_>>(),
```

Comment:
Nit: since we have the type specified in the let binding already,

```suggestion
                        Pat::OrPat(pats) => pats.pats().collect(),
```